pkg/registry/flowcontrol/flowschema/storage/storage.go

```diff
@@ -0,0 +1,113 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"context"
+	"errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/registry/generic"
+	genericregistry "k8s.io/apiserver/pkg/registry/generic/registry"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/kubernetes/pkg/apis/flowcontrol"
+	flowcontrolbootstrap "k8s.io/kubernetes/pkg/apis/flowcontrol/bootstrap"
+	"k8s.io/kubernetes/pkg/printers"
+	printersinternal "k8s.io/kubernetes/pkg/printers/internalversion"
+	printerstorage "k8s.io/kubernetes/pkg/printers/storage"
+	"k8s.io/kubernetes/pkg/registry/flowcontrol/flowschema"
+)
+
+// Container includes dummy storage for RC pods and experimental storage for Scale.
+type FlowSchemaStorage struct {
+	FlowSchema *REST
+	Status     *StatusREST
+}
+
+func NewStorage(optsGetter generic.RESTOptionsGetter) FlowSchemaStorage {
+	flowSchemaREST, flowSchemaStatusREST := NewREST(optsGetter)
+
+	return FlowSchemaStorage{
+		FlowSchema: flowSchemaREST,
+		Status:     flowSchemaStatusREST,
+	}
+}
+
+// REST implements a RESTStorage for jobs against etcd
+type REST struct {
+	*genericregistry.Store
+}
+
+// NewREST returns a RESTStorage object that will work against Jobs.
+func NewREST(optsGetter generic.RESTOptionsGetter) (*REST, *StatusREST) {
+	store := &genericregistry.Store{
+		NewFunc:                  func() runtime.Object { return &flowcontrol.FlowSchema{} },
+		NewListFunc:              func() runtime.Object { return &flowcontrol.FlowSchema{} },
+		DefaultQualifiedResource: flowcontrol.Resource("flowschemas"),
+
+		CreateStrategy: flowschema.Strategy,
+		UpdateStrategy: flowschema.Strategy,
+		DeleteStrategy: flowschema.Strategy,
+
+		// TODO: develop printer handler for flowschema resource
+		TableConvertor: printerstorage.TableConvertor{TableGenerator: printers.NewTableGenerator().With(printersinternal.AddHandlers)},
+	}
+	options := &generic.StoreOptions{RESTOptions: optsGetter}
+	if err := store.CompleteWithOptions(options); err != nil {
+		panic(err) // TODO: Propagate error up
+	}
+
+	statusStore := *store
+	statusStore.UpdateStrategy = flowschema.StatusStrategy
+
+	return &REST{store}, &StatusREST{store: &statusStore}
+}
+
+// Delete ensures that system priority classes are not deleted.
+func (r *REST) Delete(ctx context.Context, name string, deleteValidation rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	for _, pl := range flowcontrolbootstrap.SystemPriorityLevelConfigurations() {
```

Comment:
The KEP does not forbid deleting any configuration objects.  https://github.com/kubernetes/enhancements/blob/31c753ff9a5b8f9a9d89a88de67ea008e872c51d/keps/sig-api-machinery/20190228-priority-and-fairness.md#default-behavior says what happens, for any given set of configuration API objects that exist.

Reply:
in the implementation of this pull, we forbid users from creating neither `exempt` nor `globalDefault` PL objects. if we allow the deletion, they will never be recovered. wdyt?

Reply:
I think admins can create as well as delete.  If you read https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#default-behavior you will see a design that allows admins to create and delete any config objects and produces well-defined behavior at all times.

Reply:
isn't that increasing complexity? even tho the system is designed to deal with configuration changes flexibly. for now, system critical resources like service `kubernetes.default` is allowed to delete, but there's a daemon controller inside kube-apiserver to reconcile it back in an interval. if we allows deletion, we should probably do the same i.e. reconciling `system-top` and `workload-low` back

Reply:
The situation with the `kubernetes` service is more severe: the apiservers can not compensate for the absence of that service, because it is needed by clients to contact the main apiservers.  Our config objects, on the other hand, are used only inside apiservers by code that we control.  So we can easily supply default behavior when config objects are missing.  The KEP says that this happens, and it has the virtue that admins can make any changes they want.  I contemplate admins not only deleting objects but also creating alternatives.  Those two can not happen at the same instant, and the design in the KEP provides decent bridging behavior if necessary.  Allowing admins to make arbitrary changes is particularly good for an alpha feature, where us developers can not be very confident that we know the right configuration.  It is also important in scenarios with extensions to Kubernetes.

I can see an argument for making an exception for the exempt priority level.  Since it has no hair, there are no changes for admins to make to it.  But this logic applies only to that one config object.

Reply:
that's definitely true. if users are complaining with the presets, they're always welcome to make adjustment(update) the priority-level objects. as for `system-top`, they can delete related flow-schemas in order to "mute" the top priority i think?

Reply:
What I aimed for in the designed defaulting behavior is to allow administrators to make any change they want, from small isolated tweaks to wholesale replacement of all the config objects if they decide on a different organization.

In the example you mentioned, modifying all the FlowSchema objects that formerly referred to the `system-top` priority level to make them refer to other priority levels, we should not require the undesired `system-top` priority level config object to remain, right?